File: jdroid-java-firebase-database/src/main/java/com/jdroid/java/firebase/database/FirebaseCompletionListener.java
package com.jdroid.java.firebase.database;

import com.firebase.client.Firebase;
import com.firebase.client.FirebaseError;
import com.jdroid.java.exception.UnexpectedException;

public class FirebaseCompletionListener implements Firebase.CompletionListener {

	private FirebaseCountDownLatch done = new FirebaseCountDownLatch();

	@Override
	public void onComplete(FirebaseError firebaseError, Firebase firebase) {
		if (firebaseError != null) {
			done.setFirebaseException(new FirebaseException(firebaseError));
		}
		done.countDown();
	}

	public void waitOperation() {
		try {
			done.await();
			if (done.getFirebaseException() != null) {
				throw done.getFirebaseException();
			}
		} catch (InterruptedException e) {
			throw new UnexpectedException(e);
		}
	}
}

File: jdroid-java-firebase-database/src/main/java/com/dipien/firebase/database/FirebaseCompletionListener.kt
package com.dipien.firebase.database

import com.firebase.client.Firebase
import com.firebase.client.FirebaseError

class FirebaseCompletionListener : Firebase.CompletionListener {

    private val done = FirebaseCountDownLatch()

    override fun onComplete(firebaseError: FirebaseError?, firebase: Firebase) {
        if (firebaseError != null) {
            done.firebaseException = FirebaseException(firebaseError)
        }
        done.countDown()
    }

    fun waitOperation() {
        try {
            done.await()
            done.firebaseException?.let {
                throw it
            }
        } catch (e: InterruptedException) {
            throw RuntimeException(e)
        }
    }
}
